src/dune_rules/lib.ml

```diff
@@ -1179,48 +1240,21 @@ end = struct
       (res, { Resolved_select.src_fn; dst_fn = result_fn })
     in
     let open Memo.O in
-    let+ res, selects, re_exports =
-      Memo.List.fold_left deps
-        ~init:(Resolve.return [], [], Resolve.return [])
-        ~f:(fun (acc_res, acc_selects, acc_re_exports) dep ->
-          let open Memo.O in
+    let open Resolved_deps_builder in
+    let+ builder =
+      Memo.List.fold_left deps ~init:empty ~f:(fun acc dep ->
           match (dep : Lib_dep.t) with
-          | Re_export (loc, name) ->
-            let+ lib = resolve_dep db (loc, name) ~private_deps in
-            let open Resolve.O in
-            let acc_re_exports =
-              let+ lib = lib
-              and+ acc_re_exports = acc_re_exports in
-              lib :: acc_re_exports
-            in
-            let acc_res =
-              let+ lib = lib
-              and+ acc_res = acc_res in
-              lib :: acc_res
-            in
-            (acc_res, acc_selects, acc_re_exports)
-          | Direct (loc, name) ->
-            let+ lib = resolve_dep db (loc, name) ~private_deps in
-            let acc_res =
-              let open Resolve.O in
-              let+ lib = lib
-              and+ acc_res = acc_res in
-              lib :: acc_res
-            in
-            (acc_res, acc_selects, acc_re_exports)
+          | Re_export lib ->
+            let+ lib = resolve_dep db lib ~private_deps in
+            acc |> add_re_exports ~re_exports:lib |> add_resolved ~resolved:lib
```

Comment:
Same issue with `add_re_exports`. it seems wrong to call it without the appropriate `add_resolved` so it should do that work already.

Reply:
sure. I did that in a way that manipulates fields more directly, but we can do that in a way that maintains the invariant.

Reply:
The type is abstract, so it seems like the API is responsible for maintaining the invariants of it.